### ISSUE TITLE
avs-device-sdk 1.9: Alexa not able to control music from bluetooth

### ISSUE THREAD

**Issue by sumit3817** [alexa/avs-device-sdk#1212]
Hello AVS Team,

I am using 1.9 version of avs-device-sdk compiled with bluetooth flag.

I am able to pair my phone through alexa command.

But when I play music from bluetooth (Wync music app), now when I give the alexa command "Alexa play music", It doesn't pause the bluetooth music. Both music plays parallely.

Configuration:-
  All the bluetooth dependencies suggested by avs team (documented) are done.

Steps to reproduce:-
1. Run the alexa app on ubuntu environment (I am using artik board, avrcp profile is configured already)
2. Now I run these command parallely:-
      pactl unload-module module-bluetooth-discover
      pactl unload-module module-bluetooth-policy
     pactl load-module module-bluetooth-discover
     pactl load-module module-bluetooth-policy
These commands runs successfully.

3. Now I connect my phone through Alexa command " Alexa, connect my phone" . 
   This step also runs successfully and I am able to connect my phone to artik using alexa command 
    through bluetooth.
4. Now I play music from my phone on bluetooth. This also works.

5. Now I play music through alexa command.
   In this step, both the music runs parallely (bluetooth music and avs music)

Can you please suggest how to control music on device (in my case artik) from alexa??

**Comment by sumit3817** [comment]
Hello AVS Team,

I have attached the critical logs of the scenario by enabling -DACSDK_EMIT_SENSITIVE_LOGS=ON
The scanario is as follows:-
1. Run avs-device-sdk (version 1.9 compiled with bluetooth flag)  client on my device (artik)
2. Execute the following commands on the board 
      pactl unload-module module-bluetooth-discover
    pactl unload-module module-bluetooth-policy
    pactl load-module module-bluetooth-discover
    pactl load-module module-bluetooth-policy
3. Connect to my phone by saying "Alexa Connect to my phone"
4. Play music on my device through phone (using Wync music app)
5. Now play music using avs by saying "Alexa Play music"
The logs attached shows both the music are running parallely.

Please help me here to solve this problem.

[bluetooth_control_critical_logs.txt](https://github.com/alexa/avs-device-sdk/files/2836339/bluetooth_control_critical_logs.txt)

**Comment by kjkh** [comment]
Hi @sumit3817,

Please ensure that you run Step 2 only after the SDK has started. Can you try that and see if you still have the issue of music playing in parallel?

**Comment by sumit3817** [comment]
Hi,

Yes, I am running the Step2 only after the SDK has started.
Still I am facing the issue music playing in parallel.

**Comment by sumit3817** [comment]
Hello AVS Team,

Can anyone help me here?

**Comment by kjkh** [comment]
Hi @sumit3817,

In the logs I am seeing that the Bluetooth Capability Agent is not recognizing that streaming has started. Additionally, there is this error:

`2019-02-06 12:33:55.918 [  e] 5 BlueZDeviceManager:onMediaStreamPropertyChanged:reason=pathMismatch,path=/org/bluez/hci0/dev_60_9A_C1_E3_29_19/fd0,streamingDevicePath=`

These two signs typically indicate that the SDK is not handling the playback, and some other system (typically PulseAudio) is handling it instead.

Can you try running only
```
pactl unload-module module-bluetooth-discover
pactl unload-module module-bluetooth-policy
```
before starting the SDK, and then seeing if there is still music overlap? By not reloading the modules, the A2DPSource use case will not work. This is okay since we are attempting to debug A2DPSink.

**Comment by sumit3817** [comment]
Hi,
By running only 

   pactl unload-module module-bluetooth-discover
   pactl unload-module module-bluetooth-policy

before starting sdk, the music is unable to play using bluetooth.
So only alexa music is played.

Can you please suggest

**Comment by sumit3817** [comment]
Hi,
I have uploaded the bluetooth.service and bluetooth.conf file.

Please let me know anymore files needed 

[bluetooth.conf.txt](https://github.com/alexa/avs-device-sdk/files/2924382/bluetooth.conf.txt)
[bluetooth.service.txt](https://github.com/alexa/avs-device-sdk/files/2924383/bluetooth.service.txt)

**Comment by kjkh** [comment]
Hi @sumit3817,

The two config files you have attached look correct. Which platform are you using? Also in 1.12 we have added the functionality to allow for programmatic set up of BT endpoints with regards to PulseAudio. Can you try upgrading to 1.12 and following the new instructions at https://github.com/alexa/avs-device-sdk/wiki/Ubuntu-Linux-Quick-Start-Guide#include-bluetooth-optional? If there is still music overlap, please provide DEBUG9 logs.

**Comment by sumit3817** [comment]
Hi,

[bluetooth_music_overlap.txt](https://github.com/alexa/avs-device-sdk/files/2956010/bluetooth_music_overlap.txt)

I have also tried with 1.12 and have used the flag -DBLUETOOTH_BLUEZ_PULSEAUDIO_OVERRIDE_ENDPOINTS=ON

Still I am facing the same issue of music overlap.
I have attached here the sensitive logs with DEBUG9 enabled.

**Comment by kjkh** [comment]
Hi @sumit3817,

In the logs I am seeing this specific line:
`2019-03-12 09:05:23.267 [  1] 5 DefaultClient:bluetoothCapabilityAgentDisabled:reason=nullBluetoothDeviceManager`

This means that the Bluetooth CA is not enabled, and the Bluetooth streaming you're hearing is completely being handled by the platform.

What do your cmake commands look like, and have you made modifications to the SampleApp or DefaultClient?

**Comment by sumit3817** [comment]
Hi,

There are no modifications in DefaultClient but yes I am compiling SampleApp seperately.

I am cross compiling the avs-device-sdk and SampleApp seperately for artik.

Here are my cmake commands:-

**cmake command for compiling avs-device-sdk -**
cmake -DCMAKE_TOOLCHAIN_FILE=/home/sumit/magicbox_alexa_build/toolchain.cmake /home/sumit/magicbox_alexa_build/1.12/sdk-source/avs-device-sdk -DCMAKE_INSTALL_PREFIX=/home/sumit/magicbox_alexa_build/build_alexa_debug/ -DKWD=ON -DGSTREAMER_MEDIA_PLAYER=ON -DPORTAUDIO=ON -DPORTAUDIO_LIB_PATH=/home/sumit/magicbox_alexa_build/build_alexa_debug/lib/libportaudio.a -DPORTAUDIO_INCLUDE_DIR=/home/sumit/magicbox_alexa_build/build_alexa_debug/include -DAMAZONLITE_KEY_WORD_DETECTOR=ON -DAMAZONLITE_KEY_WORD_DETECTOR_LIB_PATH=/home/sumit/magicbox_alexa_build/build_alexa_debug/lib/libpryon_lite.a -DAMAZONLITE_KEY_WORD_DETECTOR_INCLUDE_DIR=/home/sumit/magicbox_alexa_build/build_alexa_debug/include -DAMAZONLITE_KEY_WORD_DETECTOR_DYNAMIC_MODEL_LOADING=ON -DBLUETOOTH_BLUEZ=ON -DCMAKE_BUILD_TYPE=DEBUG -DACSDK_EMIT_SENSITIVE_LOGS=ON -DBLUETOOTH_BLUEZ_PULSEAUDIO_OVERRIDE_ENDPOINTS=ON

**cmake command for compiling SampleApp seperately-**
cmake -DCMAKE_TOOLCHAIN_FILE=/home/sumit/magicbox_alexa_build/toolchain.cmake /home/sumit/magicbox_alexa_build/1.12/sdk-source/Simplifi_Alexa -DCMAKE_PREFIX_PATH=/home/sumit/magicbox_alexa_build/build_alexa_debug/ -DKWD=ON -DGSTREAMER_MEDIA_PLAYER=ON -DPORTAUDIO=ON -DPORTAUDIO_LIB_PATH=/home/sumit/magicbox_alexa_build/build_alexa_debug/lib/libportaudio.a -DPORTAUDIO_INCLUDE_DIR=/home/sumit/magicbox_alexa_build/build_alexa/include -DAMAZONLITE_KEY_WORD_DETECTOR=ON -DAMAZONLITE_KEY_WORD_DETECTOR_DYNAMIC_MODEL_LOADING=ON -DBLUETOOTH_BLUEZ=ON -DCMAKE_BUILD_TYPE=DEBUG -DACSDK_EMIT_SENSITIVE_LOGS=ON -DBLUETOOTH_BLUEZ_PULSEAUDIO_OVERRIDE_ENDPOINTS=ON



Please suggest if there is a need in change to compile SampleApp seperately for enabling bluetooth functionality.

**Comment by mvelegon-amzn** [comment]
Hi @sumit3817 could you confirm that you are running SDK v1.12? I don't think you should be seeing the `DefaultClient:bluetoothCapabilityAgentDisabled:reason=nullBluetoothDeviceManager` error in the logs with your cmake command. Also, please send the latest log as well.

**Comment by sumit3817** [comment]
Hi,

Yes I am running v1.12.
I have attached the latest logs.
These logs are for the basic SampleApp which is compiled with avs-device-sdk itself.
Still I am facing the music overlap issues.

I am suspecting the error in the following log line:-
**BlueZDeviceManager:onMediaStreamPropertyChanged:reason=pathMismatch,path=/org/bluez/hci0/dev_60_9A_C1_E3_29_19/fd4,streamingDevicePath=**


[sdk1.12_SampleApp_BT_Music_Overlap.txt](https://github.com/alexa/avs-device-sdk/files/3013590/sdk1.12_SampleApp_BT_Music_Overlap.txt)

**Comment by mvelegon-amzn** [comment]
Hi @sumit3817 in the previous log the Bluetooth CA was not registered and that is why we were seeing  `DefaultClient:bluetoothCapabilityAgentDisabled:reason=nullBluetoothDeviceManager` in the log.  However in the new log you sent, I do see that the Bluetooth CA is registered.

I am seeing this error in the logs: `DBusPropertiesProxy:getStringPropertyFailed:Failed to get string property=UUID,error=GDBus.Error\:org.freedesktop.DBus.Error.UnknownObject\: Method "Get" with signature "ss" on interface "org.freedesktop.DBus.Properties" doesn't exist`

In order to rule out issues with the BlueZ version you are using, can you paste the output of the following command on Ubuntu : `dpkg --status bluez | grep '^Version:'` . Please note that the minimum version of BlueZ we support is 5.37.

**Comment by sumit3817** [comment]
Hi,

Currently the version of blueZ is 5.42.

[root@artik ~]# dpkg --status bluez | grep 'Version:'
Version: 5.42-0ubuntu1.2

Please suggest.

**Comment by mvelegon-amzn** [comment]
Hi @sumit3817 looking at this issue and the one listed here: https://github.com/alexa/avs-device-sdk/issues/1265 it seems like this could be caused because of the way blueZ interacts with SDK, PulseAudio and MPlayer. 
To help us debug this issue more, could you do following:

1) Close the SampleApp instance you are running.
2) Run this command `sudo journalctl --unit=bluetooth -f`
3) Now in a new terminal tab run the SampleApp

Please post the output you see here. it should look something like this:

> Mar 29 00:36:19 uf20db51c2d8d57.ant.amazon.com bluetoothd[2675]: Endpoint registered: sender=:1.247 path=/com/amazon/alexa/sdk/sinkendpoint
> Mar 29 00:36:19 uf20db51c2d8d57.ant.amazon.com bluetoothd[2675]: Player registered: sender=:1.247 path=/org/mpris/MediaPlayer2
> Mar 29 00:36:19 uf20db51c2d8d57.ant.amazon.com bluetoothd[2675]: Endpoint unregistered: sender=:1.240 path=/MediaEndpoint/A2DPSource
> Mar 29 00:36:19 uf20db51c2d8d57.ant.amazon.com bluetoothd[2675]: Endpoint unregistered: sender=:1.240 path=/MediaEndpoint/A2DPSink
> Mar 29 00:36:19 uf20db51c2d8d57.ant.amazon.com bluetoothd[2675]: Endpoint registered: sender=:1.248 path=/MediaEndpoint/A2DPSource
> Mar 29 00:36:19 uf20db51c2d8d57.ant.amazon.com bluetoothd[2675]: Endpoint registered: sender=:1.248 path=/MediaEndpoint/A2DPSink



Thanks

**Comment by sumit3817** [comment]
Hi,

Here are the logs:-


[root@artik ~]# sudo journalctl --unit=bluetooth -f
-- Logs begin at Fri 2019-03-29 14:02:59 UTC. --
Mar 29 14:07:18 artik bluetoothd[2097]: Endpoint unregistered: sender=:1.12 path=/com/amazon/alexa/sdk/sinkendpoint
Mar 29 14:07:18 artik bluetoothd[2097]: Player unregistered: sender=:1.12 path=/org/mpris/MediaPlayer2
Mar 29 14:07:28 artik bluetoothd[2097]: Endpoint registered: sender=:1.14 path=/com/amazon/alexa/sdk/sinkendpoint
Mar 29 14:07:28 artik bluetoothd[2097]: Player registered: sender=:1.14 path=/org/mpris/MediaPlayer2
Mar 29 14:07:28 artik bluetoothd[2097]: Endpoint unregistered: sender=:1.13 path=/MediaEndpoint/A2DPSource
Mar 29 14:07:28 artik bluetoothd[2097]: Endpoint unregistered: sender=:1.13 path=/MediaEndpoint/A2DPSink
Mar 29 14:07:28 artik bluetoothd[2097]: Endpoint registered: sender=:1.15 path=/MediaEndpoint/A2DPSource
Mar 29 14:07:28 artik bluetoothd[2097]: Endpoint registered: sender=:1.15 path=/MediaEndpoint/A2DPSink
Mar 29 14:08:47 artik bluetoothd[2097]: Endpoint unregistered: sender=:1.14 path=/com/amazon/alexa/sdk/sinkendpoint
Mar 29 14:08:47 artik bluetoothd[2097]: Player unregistered: sender=:1.14 path=/org/mpris/MediaPlayer2
Mar 29 14:09:40 artik bluetoothd[2097]: Endpoint registered: sender=:1.16 path=/com/amazon/alexa/sdk/sinkendpoint
Mar 29 14:09:40 artik bluetoothd[2097]: Player registered: sender=:1.16 path=/org/mpris/MediaPlayer2
Mar 29 14:09:40 artik bluetoothd[2097]: Endpoint unregistered: sender=:1.15 path=/MediaEndpoint/A2DPSource
Mar 29 14:09:40 artik bluetoothd[2097]: Endpoint unregistered: sender=:1.15 path=/MediaEndpoint/A2DPSink
Mar 29 14:09:40 artik bluetoothd[2097]: Endpoint registered: sender=:1.17 path=/MediaEndpoint/A2DPSource
Mar 29 14:09:40 artik bluetoothd[2097]: Endpoint registered: sender=:1.17 path=/MediaEndpoint/A2DPSink

**Comment by sumit3817** [comment]
Hello Amazon Team,

As the link 
 https://github.com/alexa/avs-device-sdk/wiki/Dependencies#bluetooth-dependencies
says that minimum pulseaudio version must be 12.2 or earlier.

I check the pulseaudio version in my system. It is 12.0.

So this link mentions **minimum version of pulseaudio must be 12.2 or earlier** which is confusing.

Is the version pulseaudio 12.0 is okay for the avs-device-sdk app?

Thanks

**Comment by samcoult** [comment]
Hi @sumit3817 , the dependencies listed are for minimum recommended version. It is possible an older version of each dependency *may* work, but we cannot guarantee there will be no issues.

**Comment by samcoult** [comment]
I am going to close this ticket for inactivity, feel free to reply or open a new ticket for further issues.
